sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -136,3 +141,53 @@ private[sql] case class SparkUserDefinedFunction(
     }
   }
 }
+
+/**
+ * A User Defined Function for wrapping an [[Aggregator]] in a form consumable by Data Frames.
+ * Can be either instantiated directly or via registration, for example:
+ * {{{
+ *   val agg: Aggregator[IN, BUF, OUT] = // typed aggregator
+ *   val udaf1 = UserDefinedAggregator(agg)
+ *   val udaf2 = spark.udf.registerAggregator("agg", agg)
+ * }}}
+ *
+ * @since 3.0.0
+ */
+@Experimental
+case class UserDefinedAggregator[IN: TypeTag, BUF, OUT](
```

Comment:
This should be `private[sql]` like `SparkUserDefinedFunction`. It's really painful to expose a case class to end-users, as it exposes a lot of stuff like the `apply`, `unapply`, `copy` methods.